products/libllbuild/BuildSystem-C-API.cpp

```diff
@@ -292,6 +292,20 @@ class CAPIBuildSystemFrontendDelegate : public BuildSystemFrontendDelegate {
   void cancel() override {
     BuildSystemFrontendDelegate::cancel();
   }
+
+  virtual void cycleDetected(const std::vector<core::Rule*>& items) override {
+    llb_data_t* rules = (llb_data_t*)malloc(items.size() * sizeof(llb_data_t));
+    int idx = 0;
+    for (std::vector<core::Rule*>::const_iterator it = items.begin(); it != items.end(); ++it) {
+      core::Rule* rule = *it;
+      llb_data_t ruleName{ rule->key.size(), (const uint8_t*) rule->key.data() };
+      rules[idx++] = ruleName;
+    }
+
+    cAPIDelegate.cycleDetected(cAPIDelegate.context, rules, items.size());
```

Comment:
I don't think this is quite good enough. The problem is that this is exposing the raw encoded key names, and the client _a priori_ has no way to decode those. We are going to need to end up exposing some parts of the BuildKey type through the C API, I think -- the most minimal form would we start by returning node names and command names (the things the client knows about), and for the other things we return something less usable (like a type code + string description).